src/nnbench/reporter/file.py

```diff
@@ -0,0 +1,205 @@
+            # Implement your custom parsing logic here
+            ...
+        def write_records(self, records):
+            # Implement your custom file writing logic here
+            ...
+    # Register your custom parser with a distinct file type
+    MyCustomParser.register("my_custom_format")
+    # Usage:
+    new_record = ...  # Load records in your custom format
+    append_record_to_records(records, new_record, "my_custom_format")
+    ```
+    """
+
+    def parse_file(self, records: str) -> Any:
+        """Parses records and returns a list of parsed data.
+
+        Args:
+        -----
+            `records:` A list or iterator of record strings.
+
+        Returns:
+        --------
+            A list of parsed records.
+        """
+        raise NotImplementedError
+
+    def write_records(self, records: Any[BenchmarkRecord], record: BenchmarkRecord) -> str:
+        """Appends a record to the existing records based on the file type.
+
+        Args:
+        -----
+            `records:` A list of parsed records.
+            `record:` The record string to append.
+            `file_type:` The file type (string).
+
+        Returns:
+        --------
+            A string form of the content to be written in a file.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def register(cls, file_type: str) -> None:
+        """Registers a parser for a specific file type.
+
+        Args:
+            `file_type:` The file type (string)
+        """
+        parsers[file_type] = cls
+
+    @staticmethod
+    def get_parser(file_type: str):
+        """Gets the registered parser for a file type.
+
+        Args:
+            `file_type:` The file type (string)
+
+        Returns:
+        --------
+            The registered RecordParser, or None if not found.
+        """
+        return parsers.get(file_type)
+
+
+class JsonParser(Parser):
+    def parse_file(self, records: str) -> List[dict]:
+        import json
+
+        try:
+            return json.loads(records if records else "[]")
+        except json.JSONDecodeError:
+            raise ValueError("Unexpected records passed")
+
+    def write_records(
+        self, parsed_records: Any[BenchmarkRecord] | None, record: BenchmarkRecord
+    ) -> str:
+        import json
+
+        parsed_records.append(record)
+        return json.dumps(parsed_records)
+
+
+class YamlParser(Parser):
+    def parse_file(self, records: str) -> List[dict]:
+        import yaml
+
+        return yaml.safe_load(records) if records else []
+
+    def write_records(
+        self, parsed_records: Any[BenchmarkRecord] | None, record: BenchmarkRecord
+    ) -> str:
+        import yaml
+
+        parsed_records.append(record)
+        for element in record["benchmarks"]:
+            element["value"] = float(element["value"])
+        return yaml.dump(parsed_records)
+
+
+# Register custom parsers here
+parsers = {"json": JsonParser, "yaml": YamlParser}
```

Comment:
I like the idea of making it a module-wide default, but we should take care that

a) the variable is private (i.e. has a leading underscore) to prevent accidental export, and 
b) we should make the value structure of the map as easy as possible.

To address b), I would start by making it a `tuple[ser, de]` where `ser` is a `Callable[[IO, dict[str, Any]], None]`, i.e., a function taking a file descriptor in write mode and a record and writing it to a file, and `de` being a `Callable[[IO], dict[str, Any]]`, a function taking a file descriptor in read mode and returning the loaded record.

You can then register the SerDe factories based on whether the necessary packages are installed (json is available out of the box, yaml and toml are not).

Reply:
Should I use a class-based approach to register the SerDe factories (like I used in Parser class) ? 
One other option is to use a simple register method which takes 3 arguments (i.e., Ser and De functions and a file_type)  as arguments.
Also, I think the class-based approach is much more concise to define the ser and de methods on the user side.

Reply:
I'm happy with either, though the optional import part (i.e., erroring if `toml/yaml` are not installed) will be a bit easier in the class case.

For now, I think the quickest way is a functional approach, though. Maybe like this:

```python
_file_loaders: dict[str, tuple[Any, Any]] = {}

def yaml_load(fp: IO, options=None):
    try:
        import yaml
    except ImportError:
        raise ModuleNotFoundError("`pyyaml` is not installed")
    
    # takes no options, but the slot is useful for passing options to file loaders.
    obj = yaml.safe_load(fp)
    return BenchmarkRecord(context=obj["context"], benchmarks=obj["benchmarks"])

def yaml_save(record: BenchmarkRecord, fp: IO, options=None) -> None:
    try:
        import yaml
    except ImportError:
        raise ModuleNotFoundError("`pyyaml` is not installed")

    yaml.safe_dump(record, fp, **(options or {})


_file_loaders["yaml"] = (yaml_save, yaml_load)
```

With an option of defining e.g. a `register_file_io(ser, de)` later to do the dict insertion if we want.